python/tests/test_tool.py

```
@@ -68,13 +68,39 @@ def upper(s: str) -> str:
+
+def test_tool_without_effect_keeps_metadata_unspecified() -> None:
+    @tool
+    def apply(value: str) -> str:
+        return value
+
+    assert get_tool_metadata(apply).effect is None
+
+
+@pytest.mark.parametrize("effect", ["invalid", "delete", "READ"])
+def test_tool_rejects_invalid_effect(effect: str) -> None:
+    with pytest.raises(ValueError, match="read.*write.*update.*destructive"):
```

Ruff (0.16.3)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)